dask/bag/tests/test_bag.py

```diff
@@ -1581,3 +1582,35 @@ def test_dask_layers_to_delayed():
     assert i.dask.layers.keys() == {"delayed-" + name}
     assert i.dask.dependencies == {"delayed-" + name: set()}
     assert i.__dask_layers__() == ("delayed-" + name,)
+
+
+def test_to_dataframe_optimize_graph():
+    x = db.from_sequence(
+        [{"name": "test1", "v1": 1}, {"name": "test2", "v1": 2}], npartitions=2
+    )
+
+    # linear operations will be fused by graph optimization
+    with dask.annotate(foo=True):
+        y = x.map(lambda a: dict(**a, v2=a["v1"] + 1))
+        y = y.map(lambda a: dict(**a, v3=a["v2"] + 1))
+
+    # verifying the maps are not fused yet
+    assert len(y.dask) > y.npartitions
+
+    # with optimizations
+    d = y.to_dataframe()
+
+    # Graph has been fused, so there are three tasks per partition
+    assert len(d.dask) == d.npartitions * 3
```

Comment:
Why three tasks per partition? I'd expect just one task per partition (something like`from_sequence-lambda-lambda-to_dataframe`)

Reply:
I see. That doesn't seem to be happening.

Graph keys with optimization:
```
[('to_dataframe-73ea50b...0b0d1ea42a', 0), 
('to_dataframe-73ea50b...0b0d1ea42a', 1), 
('from_sequence-5a9b7b...6fae68202a', 0), 
('from_sequence-5a9b7b...6fae68202a', 1), 
('lambda-to_dataframe-...0b0d1ea42a', 1), 
('lambda-to_dataframe-...0b0d1ea42a', 0)]
```

Graph keys without optimization:
```
[('from_sequence-5a9b7b...6fae68202a', 0), 
('from_sequence-5a9b7b...6fae68202a', 1), 
('lambda-61f6841c3d3ee...968754d461', 0),
('lambda-61f6841c3d3ee...968754d461', 1), 
('lambda-870bd6672f293...50b9b0a0dd', 0),
('lambda-870bd6672f293...50b9b0a0dd', 1),
('to_dataframe-73ea50b...0b0d1ea42a', 0), 
('to_dataframe-73ea50b...0b0d1ea42a', 1)]
```

There is some optimization happening as you can see. Perhaps there could be a simpler test case that would make the difference with and without optimization clearer?

Reply:
Ok I looked into this, and
1. The `from_sequence` tasks don't get fused into the `map` tasks for whatever reason (this is probably some obscure policy of the `fuse` optimization)
2. All the lambdas and the `to_dataframe` get fused as we expect
3. Since `fuse` changes the task names, it also re-inserts the final output keys, which are just aliases to the fused keys:
    ```
    ('to_dataframe-a0543e6c6e4d40b987b5d4c131a9a32d', 0):('lambda-to_dataframe-...c131a9a32d', 0)
    ('to_dataframe-a0543e6c6e4d40b987b5d4c131a9a32d', 1):('lambda-to_dataframe-...c131a9a32d', 1)
    ```

So it's all working fine, just hard to explain why it's `d.npartitions * 3`. The most readable thing might actually be
```diff
diff --git a/dask/bag/tests/test_bag.py b/dask/bag/tests/test_bag.py
index ed148ba1..b0dba020 100644
--- a/dask/bag/tests/test_bag.py
+++ b/dask/bag/tests/test_bag.py
@@ -1626,15 +1626,16 @@ def test_to_dataframe_optimize_graph():
     with dask.annotate(foo=True):
         y = x.map(lambda a: dict(**a, v2=a["v1"] + 1))
         y = y.map(lambda a: dict(**a, v3=a["v2"] + 1))
+        y = y.map(lambda a: dict(**a, v3=a["v2"] + 1))
 
     # verifying the maps are not fused yet
-    assert len(y.dask) > y.npartitions
+    assert len(y.dask) == y.npartitions * 4
 
     # with optimizations
     d = y.to_dataframe()
 
-    # Graph has been fused, so there are three tasks per partition
-    assert len(d.dask) == d.npartitions * 3
+    # All the `map` tasks have been fused
+    assert len(d.dask) < len(y.dask)
 
     # no optimizations
     d2 = y.to_dataframe(optimize_graph=False)
```